### PR TITLE
minimum and maximum length for text fields and autosize for textarea

### PR DESCRIPTION
**Pull request by seryozhabaleyko**

### CODE REVIEW

File: src/components/BaseTextArea/index.vue [hunk 1]
```diff
@@ -1,25 +1,115 @@
...
+
+export default defineComponent<Props>({
   name: 'BaseTextArea',
   props: {
-    value: String,
+    value: {
+      type: [String, Number],
+      default: '',
+    },
+    resize: {
```
Comment (saranchonkau) [inline src/components/BaseTextArea/index.vue]: we can remove it

File: src/components/BaseTextArea/index.vue [hunk 2]
```diff
@@ -1,25 +1,115 @@
...
+      },
+    }));
+
+    const resizeTextarea = () => {
+      const { autosize, rows } = props;
+
+      if (autosize) {
+        const minRows = typeof autosize === 'object' ? autosize.minRows : rows;
+        const maxRows = typeof autosize === 'object' ? autosize.maxRows : null;
+        _textareaCalcStyle.value = calcTextareaHeight(
```
Comment (saranchonkau) [inline src/components/BaseTextArea/index.vue]: use `autosize` package

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+      v-bind="$attrs"
+      v-on="$listeners"
+    />
+
+    <progress
+      class="progress"
+      max="100"
+      :value="percent"
+      :class="[{ 'lower-limit': isLowerLimit, 'upper-limit': isUpperLimit }]"
+    />
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: reimplement via `div/span`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+  },
+  inheritAttrs: false,
+  props: {
+    name: {
+      type: String,
+      required: true,
+    },
+    label: String,
+    value: {
+      type: [String, Number],
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: remove `Number`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        : minLength;
+    });
+
+    const upperLimit = computed<number>(() => {
+      const { maxLength } = props;
+
+      return typeof maxLength === 'string'
+        ? parseInt(maxLength, 10)
+        : maxLength;
+    });
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: remove

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+      return (modelValue.value || '').length;
+    });
+
+    const lowerLimit = computed<number>(() => {
+      const { minLength } = props;
+
+      return typeof minLength === 'string'
+        ? parseInt(minLength, 10)
+        : minLength;
+    });
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: remove

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+          'email',
+          'password',
+          'date',
+          'time',
+          'textarea',
+        ].includes(value);
+      },
+    },
+    error: String,
+    minLength: [String, Number],
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: make `Number`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+      return typeof maxLength === 'string'
+        ? parseInt(maxLength, 10)
+        : maxLength;
+    });
+
+    const isLowerLimit = computed<boolean>(() => {
+      return textLength.value < lowerLimit.value;
+    });
+
+    const isUpperLimit = computed<boolean>(() => {
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: `isHigherThanMaxLength`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+    const upperLimit = computed<number>(() => {
+      const { maxLength } = props;
+
+      return typeof maxLength === 'string'
+        ? parseInt(maxLength, 10)
+        : maxLength;
+    });
+
+    const isLowerLimit = computed<boolean>(() => {
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: `isLowerThanMinLength`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  setup(props: Props) {
+    const modelValue = ref<string | number>(props.value);
+
+    const textLength = computed<number>(() => {
+      if (typeof modelValue.value === 'number') {
+        return String(modelValue.value).length;
+      }
+
+      return (modelValue.value || '').length;
+    });
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: use `props.value.length`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+          transform: 'translateX(-50%)',
+        }
+      } else {
+        obj = {
+          right: 0,
+        }
+      }
+
+      return obj;
+    });
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: replace with ternary operator or one `if` condition

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 12]
```diff
@@ -0,0 +1,234 @@
+<template>
+  <FormGroup>
+    <InputLabel v-if="Boolean(label)" :for="name">{{ label }}</InputLabel>
+
+    <BaseInput
+      v-if="type !== 'textarea'"
+      :id="name"
+      v-model="modelValue"
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: remove `v-model`

File: src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+      :type="type"
+      :value="value"
+      v-bind="$attrs"
+      v-on="$listeners"
+    />
+
+    <BaseTextArea
+      v-else
+      :id="name"
+      v-model="modelValue"
```
Comment (saranchonkau) [inline src/components/FormFieldMinMaxLength/FormFieldMinMaxLength.vue]: remove `v-model`

File: package.json [hunk 14]
```diff
@@ -23,6 +23,8 @@
     "@babel/runtime": "^7.11.2",
     "@ckeditor/ckeditor5-vue": "^1.0.2",
     "@tager/admin-wysiwyg": "^0.1.0",
+    "@types/autosize": "^3.0.7",
```
Comment (saranchonkau) [inline package.json]: move it to devDependencies

File: src/components/BaseTextArea/BaseTextArea.vue [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+} from '@vue/composition-api';
+import autosize from 'autosize';
+import { Nullable } from '@tager/admin-services';
+
+type ResizePropType = 'none' | 'both' | 'horizontal' | 'vertical';
+type AutocompletePropType = 'on' | 'of';
+
+interface Props {
+  value: string;
+  resize: ResizePropType;
```
Comment (saranchonkau) [inline src/components/BaseTextArea/BaseTextArea.vue]: remove `resize` from props

File: src/components/BaseTextArea/BaseTextArea.vue [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+import {
+  computed,
+  defineComponent,
+  onMounted,
+  ref,
+} from '@vue/composition-api';
+import autosize from 'autosize';
+import { Nullable } from '@tager/admin-services';
+
+type ResizePropType = 'none' | 'both' | 'horizontal' | 'vertical';
```
Comment (saranchonkau) [inline src/components/BaseTextArea/BaseTextArea.vue]: remove `ResizePropType `

File: src/components/BaseTextArea/BaseTextArea.vue [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+type AutocompletePropType = 'on' | 'of';
+
+interface Props {
+  value: string;
+  resize: ResizePropType;
+  autosize: boolean;
+  rows: number;
+  autocomplete: AutocompletePropType;
```
Comment (saranchonkau) [inline src/components/BaseTextArea/BaseTextArea.vue]: use directly `'on' | 'of'` and remove `AutocompletePropType`

File: src/components/BaseTextArea/BaseTextArea.vue [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        context.emit('change', value);
+      },
+    }));
+
+    onMounted(() => {
+      if (props.autosize && textareaRef.value) {
+        autosize(textareaRef.value);
+      }
+    });
+
```
Comment (saranchonkau) [inline src/components/BaseTextArea/BaseTextArea.vue:72]: run `autosize.destroy` in `onUnmounted` hook

https://composition-api.vuejs.org/api.html#lifecycle-hooks
https://github.com/jackmoore/autosize/blob/master/src/autosize.js#L242

File: src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+          'password',
+          'date',
+          'time',
+          'textarea',
+        ].includes(value);
+      },
+    },
+    error: String,
+    minLength: Number,
+    maxLength: Number,
```
Comment (saranchonkau) [inline src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue]: How can we pass `maxLength` to input, if we already have prop with such name here?

File: src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+    const isLowerThanMinLength = computed<boolean>(() => {
+      return props.value.length < props.minLength;
+    });
+
+    const isHigherThanMaxLength = computed<boolean>(() => {
+      return props.value.length > props.maxLength;
+    });
+
+    const percent = computed<number>(() => {
+      const A1: number = props.value.length;
```
Comment (saranchonkau) [inline src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue]: what does `A1` means?)

File: src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+      const A2: number = props.maxLength ? props.maxLength : props.minLength;
+      return (A1 / A2) * 100;
+    });
+
+    const color = computed<string>(() => {
```
Comment (saranchonkau) [inline src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue:123]: add comment here
`// TODO Refactor it. Maybe move this logic to Progress component`

File: src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue (hunk 19 above)
Comment (seryozhabaleyko) [reply src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue]: not maxLength, but maxlength
image

File: src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue (hunk 20 above)
Comment (seryozhabaleyko) [reply src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue]: I thought for a very long time and did not come up with anything better like A1 and A2.

Comment (seryozhabaleyko) [reply src/components/FormFieldRecommendedLengthInput/FormFieldRecommendedLengthInput.vue]: image

**Pull request merged**